File: Divisão e Conquista/G.java
import java.io.*;
import java.util.*;

class G{
    public static List<Long> fib = new ArrayList<>();
    public static void calculaFibs(){
        fib.add(1L);
        fib.add(1L);
        while(true){
            long proximo = fib.get(fib.size() - 1) + fib.get(fib.size() - 2);
            fib.add(proximo);
            if(proximo > 1e18) break;
        }
    }


    public static char palavraFib(long n){
        if(n == 0) return 'b';
        if(n == 1) return 'a';

        long fibAux = 1;
        for(int i = fib.size() - 1; i >= 0; i--){
            if(fib.get(i) <= n){
                fibAux = fib.get(i);
                break;
            }
        }

        return palavraFib(n - fibAux);
    }

    public static void main(String[] args) throws Exception{
        calculaFibs();

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        for(int i = 0; i < n; i++){
            long aux = Long.parseLong(br.readLine());
            System.out.println(palavraFib(aux));
        }
    }
}

File: Divisão e Conquista/G.cpp
#include <bits/stdc++.h>

using namespace std;

using llu = long long unsigned int;
vector<llu> fib;
void calculaFibs(){
    fib.push_back(1);
    fib.push_back(1);
    while(true){
        llu proximo = fib[fib.size() - 1] + fib[fib.size() - 2];
        fib.push_back(proximo);
        if(proximo > 1e18) break;
    }
}

char palavraFib(long n){
    if(n == 0) return 'b';
    if(n == 1) return 'a';

    long fibAux = 1;
    for(int i = fib.size() - 1; i >= 0; i--){
        if(fib[i] <= n){
            fibAux = fib[i];
            break;
        }
    }

    return palavraFib(n - fibAux);
}

int main(){
    calculaFibs();

    int n;
    cin >> n;

    for(int i = 0; i < n; i++){
        llu aux;
        cin >> aux;
        cout << palavraFib(aux) << "\n";
    }
}
